### PR TITLE
Add foreman-tasks manual

### PR DESCRIPTION
**Pull request by ares**
@iNecas or @adamruzicka could you please check that I'm not making up something? :-)

### CODE REVIEW

File: plugins/foreman_tasks/0.8/index.md
Comment (adamruzicka) [inline plugins/foreman_tasks/0.8/index.md]: s/reoccuring/recurring

File: plugins/foreman_tasks/0.8/index.md
Comment (adamruzicka) [inline plugins/foreman_tasks/0.8/index.md]: s/impotring/importing

File: plugins/foreman_tasks/0.8/index.md
Comment (adamruzicka) [inline plugins/foreman_tasks/0.8/index.md]: s/actuall/actually

**Review by adamruzicka — CHANGES_REQUESTED** [review]
Very well written, +1. However I noticed some small typos in the second half

**Comment by ares** [comment]
thanks @adamruzicka I fixed what you commented on

File: plugins/foreman_tasks/0.8/index.md
Comment (ohadlevy) [inline plugins/foreman_tasks/0.8/index.md]: Why do I see .. ?

File: plugins/foreman_tasks/0.8/index.md
Comment (ohadlevy) [inline plugins/foreman_tasks/0.8/index.md]: IMHO swap the condition, e.g.
an error has occurred and it was programmed to pause on error

File: plugins/foreman_tasks/0.8/index.md
Comment (ohadlevy) [inline plugins/foreman_tasks/0.8/index.md]: can you add examples of why you wont know it?

File: plugins/foreman_tasks/0.8/index.md
Comment (ohadlevy) [inline plugins/foreman_tasks/0.8/index.md]: then what is the difference to the above statement about pause? as you just said its mostly paused due to an error, which other state exists and why?

Comment (ares) [reply plugins/foreman_tasks/0.8/index.md]: I think the task might be paused even if no error occurred so you would see paused task with success state. @iNecas or @adamruzicka do you have any example? I think "Listen on candlepin events" was such task but I think the state was renamed there to be less confusing? The generic use case is that the task triggers some external event and pauses itself. Pause actually means it's waiting to be resumed by some external system, e.g. after the external event finished.

Anyway this Q is about specific combination of paused task with error result. As it says, in task detail, we denote this task result as pending since the task hasn't finished yet, but in task list we display the "current status". I'd prefer to always display pending until the task is stopped but it was not accepted. The reason was that this information might be useful even before the task finished.

**Comment by ares** [comment]
@ohadlevy I tried to explain what I meant and updated the rest, could you please take another look if it makes more sense now?

**Comment by ares** [comment]
@ohadlevy @adamruzicka ping

File: plugins/foreman_tasks/0.8/index.md
Comment (dLobatog) [inline plugins/foreman_tasks/0.8/index.md]: The link is wrongly formatted here. I would rather recommend using the ActiveJob Dynflow interface instead, which is fully documented by Rails and has examples all over. There's nothing special we do, except that I tweaked it to accept `def humanized_*` methods to play well with tasks.

Comment (dLobatog) [inline plugins/foreman_tasks/0.8/index.md]: The link is wrongly formatted here. I would rather recommend using the ActiveJob Dynflow interface instead, which is fully documented by Rails and has examples all over. There's nothing special we do, except that I tweaked it to accept `def humanized_*` methods to play well with tasks.

File: plugins/foreman_tasks/0.8/index.md
Comment (adamruzicka) [reply plugins/foreman_tasks/0.8/index.md]: The "Listen on candlepin events" task should always be in the `running` state.

It sounds like you're confusing tasks' `paused` state with Dynflow steps' `suspended`. A step can be `suspended` which means it is waiting for an external event. When a task/execution plan is `paused` it means it encountered an error which it can't solve on its own and needs the user to either cancel it or fix the issue and resume it.

Also `s/stoped/stopped/`

Comment (ares) [reply plugins/foreman_tasks/0.8/index.md]: Ok, thanks for clarification. I updated Q&A section to mention only error as a cause for task to be paused

File: plugins/foreman_tasks/0.8/index.md
Comment (ares) [reply plugins/foreman_tasks/0.8/index.md]: Thanks, fixed. I think ActiveJob way should be mentioned here but I think there are still cases where you want to use advanced features of foreman-tasks, such as locking, that requires you to write the task manually. So I think the developer guide still makes sense covering stuff like locking, humanized_output etc. Anyway, I'd appreciate if you put together examples with ActiveJob API to this manual once it's ready.

**Comment by ares** [comment]
I updated the PR, please take another look. I think the manual can be continuously improved and this could hopefully be already a baseline for anyone who'd like to contribute with even single sentence.

File: plugins/foreman_tasks/0.8/index.md
Comment (adamruzicka) [inline plugins/foreman_tasks/0.8/index.md]: missing period at EOL

File: plugins/foreman_tasks/0.8/index.md
Comment (adamruzicka) [inline plugins/foreman_tasks/0.8/index.md]: `s/uniq/unique/`

File: plugins/foreman_tasks/0.8/index.md
Comment (adamruzicka) [inline plugins/foreman_tasks/0.8/index.md]: Too much `debugging` on this line I'd say

File: plugins/foreman_tasks/0.8/index.md
Comment (adamruzicka) [inline plugins/foreman_tasks/0.8/index.md]: > These data contains all tasks there are present in database.

This isn't exactly true. Running just `foreman-rake foreman_tasks:export_tasks` will export all the tasks started in the last 7 days and all unsuccessful ones started in last 60 days (see [here](https://github.com/theforeman/foreman-tasks/blob/master/lib/foreman_tasks/tasks/export_tasks.rake#L231))

File: plugins/foreman_tasks/0.8/index.md
Comment (adamruzicka) [inline plugins/foreman_tasks/0.8/index.md]: `s/, ended_at/ and ended_at/` would sound more natural

**Comment by adamruzicka** [comment]
Did another read through, left some inline comments

**Comment by ares** [comment]
Thanks @adamruzicka, please take another look, all comments should be fixed. Also after this is merged, we should probably copy the manual to 0.9 and do following updates there.

**Comment by adamruzicka** [comment]
Looks good to me

**Comment by adamruzicka** [comment]
Could anyone please merge this? I'd do it myself but I don't have write access to this repository.

**Comment by iNecas** [comment]
I don't have either, perhaps @dLobatog could help?

**Comment by mmoll** [comment]
merged, díky všetkým!

**Comment by lzap** [comment]
Excellent work! Kudos.

**Pull request merged**
